plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c

```diff
@@ -0,0 +1,374 @@
+	dstr_copy(&path, path_str);
+	dstr_replace(&path, "{stream_key}", stream_key);
+	dstr_init(&stream->muxer_settings);
+	dstr_catf(&stream->muxer_settings,
+		  "http_user_agent=libobs/%s method=PUT http_persistent=1",
+		  OBS_VERSION);
+	dstr_catf(&stream->muxer_settings, " ignore_io_errors=1");
+
+	vencoder = obs_output_get_video_encoder(stream->output);
+	settings = obs_encoder_get_settings(vencoder);
+	keyint_sec = obs_data_get_int(settings, "keyint_sec");
+	if (keyint_sec)
+		dstr_catf(&stream->muxer_settings, " hls_time=%d", keyint_sec);
+
+	obs_data_release(settings);
+
+	start_pipe(stream, path.array);
+	dstr_free(&path);
+
+	if (!stream->pipe) {
+		obs_output_set_last_error(
+			stream->output, obs_module_text("HelperProcessFailed"));
+		warn("Failed to create process pipe");
+		return false;
+	}
+
+	/* write headers and start capture */
+	os_atomic_set_bool(&stream->active, true);
+	os_atomic_set_bool(&stream->capturing, true);
+	os_atomic_set_bool(&stream->threading_buffer, true);
+	stream->total_bytes = 0;
+	stream->dropped_frames = 0;
+	stream->min_priority = 0;
+
+	obs_output_begin_data_capture(stream->output, 0);
+
+	dstr_copy(&stream->printable_path, path_str);
+	info("Writing to path '%s'...", stream->printable_path.array);
+
+	stream->mux_thread_joinable = pthread_create(&stream->mux_thread, NULL,
+						     write_thread, stream) == 0;
+	return (stream->mux_thread_joinable);
+}
+
+static bool write_packet_to_array(struct ffmpeg_muxer *stream,
+				  struct encoder_packet *packet)
+{
+	da_push_back(stream->mux_packets, packet);
+	return true;
+}
+
+static void drop_frames(struct ffmpeg_muxer *stream, int highest_priority)
+{
+	int num_frames_dropped = 0;
+	int i = 0;
+	while (i < stream->mux_packets.num) {
+		struct encoder_packet *packet;
+		packet = &stream->mux_packets.array[i];
+
+		/* do not drop audio data or video keyframes */
+		if (!(packet->type == OBS_ENCODER_AUDIO ||
+		      packet->drop_priority >= highest_priority)) {
+			num_frames_dropped++;
+			obs_encoder_packet_release(packet);
+			da_erase_item(stream->mux_packets, packet);
+			stream->dropped_frames++;
+			i--;
+		}
+		i++;
+	}
+
+	if (stream->min_priority < highest_priority) {
+		stream->min_priority = highest_priority;
+	}
+	if (!num_frames_dropped)
+		return;
+
+	stream->dropped_frames += num_frames_dropped;
+}
+
+static bool find_first_video_packet(struct ffmpeg_muxer *stream,
+				    struct encoder_packet *first)
+{
+	for (size_t i = 0; i < stream->mux_packets.num; i++) {
+		struct encoder_packet *cur = &stream->mux_packets.array[i];
+		if (cur->type == OBS_ENCODER_VIDEO && !cur->keyframe) {
+			*first = *cur;
+			return true;
+		}
+	}
+	return false;
+}
+
+void check_to_drop_frames(struct ffmpeg_muxer *stream, bool pframes)
+{
+	struct encoder_packet first;
+	int64_t buffer_duration_usec;
+	int priority = pframes ? OBS_NAL_PRIORITY_HIGHEST
+			       : OBS_NAL_PRIORITY_HIGH;
+	int64_t drop_threshold = 2 * stream->keyframes;
```

Comment:
drop threshold should be set based on key frame interval not keyframes. keyframes is number of keyframes. You need to add another field to the ffmpeg_muxer struct to track the key frame interval.